src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/LogicalExpression.java

```diff
@@ -112,10 +90,41 @@ public void addComparisonExpressions(List<ComparisonExpression> comparisonExpres
     }
   }
 
+  /**
+   * Add childLogicalExpression after it is been flipped with $not operator Will not increment the
+   * counter to avoid double count
+   *
+   * @param logicalExpression logicalExpression
+   */
+  public void addLogicalExpressionFlipped(LogicalExpression logicalExpression) {
+    // skip empty logic expression
+    if (logicalExpression.isEmpty()) {
+      return;
+    }
+    logicalExpressions.add(logicalExpression);
+  }
+
+  /**
+   * Add childComparisonExpressions after they have been flipped with $not operator Will not
+   * increment the counter to avoid double count
+   *
+   * @param comparisonExpressionList comparisonExpressionList
+   */
+  public void addComparisonExpressionsFlipped(List<ComparisonExpression> comparisonExpressionList) {
+    if (comparisonExpressionList.isEmpty()) {
+      return;
+    }
+    comparisonExpressions.addAll(comparisonExpressionList);
+  }
+
   public LogicalOperator getLogicalRelation() {
     return logicalRelation;
   }
 
+  public void setLogicalRelation(LogicalOperator logicalRelation) {
+    this.logicalRelation = logicalRelation;
+  }
+
   public int getTotalComparisonExpressionCount() {
```

Comment:
Why do we needs these counts?

Reply:
there are multiple usages,
like maximum filter item in the filter clause, also we need this count when we do filter capture, etc.

Before we involve the logicalExpression to give the $and $or ability, we only have one level of comparisonVariable, so this count is easy to get. But now, we have this 'tree' structure, need a counter to know how many comparisonExpression in the tree, so we don't have to traverse and count again every time we need it.